In [4]:
import numpy as np

# Points as given in the problem
points = np.array([
    [0.1, 0.6],  # P1
    [0.15, 0.71],  # P2
    [0.08, 0.9],  # P3
    [0.16, 0.85],  # P4
    [0.2, 0.3],  # P5
    [0.25, 0.5],  # P6
    [0.24, 0.1],  # P7
    [0.3, 0.2]   # P8
])


# Initial centroids
m1 = points[0]  # P1
m2 = points[7]  # P8

# Function to calculate Euclidean distance
def euclidean_distance(point, centroid):
    return np.sqrt(np.sum((point - centroid) ** 2))
# Perform one iteration of k-means
def k_means_iteration(points, m1, m2):
    cluster1, cluster2 = [], []
    for point in points:
        # Calculate distances to centroids
        dist_to_m1 = euclidean_distance(point, m1)
        dist_to_m2 = euclidean_distance(point, m2)
        # Assign point to the nearest cluster
        if dist_to_m1 < dist_to_m2:
            cluster1.append(point)
        else:
            cluster2.append(point)
    # Update centroids
    new_m1 = np.mean(cluster1, axis=0) if cluster1 else m1
    new_m2 = np.mean(cluster2, axis=0) if cluster2 else m2
    return np.array(cluster1), np.array(cluster2), new_m1, new_m2
# Run one iteration
cluster1, cluster2, new_m1, new_m2 = k_means_iteration(points, m1, m2)


# Results
print("Cluster 1 (around m1):", cluster1)
print("Cluster 2 (around m2):", cluster2)
print("Updated centroid m1:", new_m1)
print("Updated centroid m2:", new_m2)

Cluster 1 (around m1): [[0.1  0.6 ]
 [0.15 0.71]
 [0.08 0.9 ]
 [0.16 0.85]
 [0.25 0.5 ]]
Cluster 2 (around m2): [[0.2  0.3 ]
 [0.24 0.1 ]
 [0.3  0.2 ]]
Updated centroid m1: [0.148 0.712]
Updated centroid m2: [0.24666667 0.2       ]


  #np.mean(cluster1, axis=0):

Calculates the mean of all points in cluster1.
xis=0 means the mean is computed column-wise (separately for x and y coordinates).

if cluster1 else m1:
Checks if cluster1 is non-empty.
If cluster1 contains points, the centroid is updated to the mean of these points.
If cluster1 is empty (no points are assigned to the cluster), the centroid remains unchanged (m1).


#Questions based on above results

#1.Which cluster does P6[0.25,0.5] belong to?
Ans=P6 belongs to cluster m1
#2. What is the population of a cluster around m2?
Ans=3 (i.e p5,p7,p8 three data points are present in cluster m2)
#3. What is the updated value of m1 and m2?
Ans=From above resuls,
Updated centroid m1: [0.148, 0.712]
Updated centroid m2: [0.24666667, 0.2]